File: halatuju-web/src/lib/admin-api/programmes.ts
/**
 * Gift programmes and their intake years (Sabah S2b): what a programme asks for, the apply
 * copy a student reads, and the cohort years it runs.
 */
import { adminFetch, adminMutate } from './client'
import type { ApiOptions } from './client'

// ── Gift programmes and their intake years (Sabah S2b, 2026-09-02) ──────────────────────────────
//
// The screens an `org_admin` uses to stand up a gift without an engineer. Every endpoint is fenced
// server-side on the programme's own organisation; cross-tenant is 404, never 403. Nothing here is
// an access check — it is the org fence's own answer, rendered.

/** The requirement columns the tick boxes write. `null` means the test is NOT APPLIED (S2a) —
 *  the value IS the switch, so unticking is writing null and there is no companion boolean. */
export interface ProgrammeRequirements {
  min_spm_a_count: number | null
  /** The TOTAL strong count, not the extra beyond the A's: "4 A− plus 1 more at B+" is stored 5.
   *  The screen shows the difference, because that is how the rule is set and read. */
  min_spm_bplus_count: number | null
  min_stpm_pngk: number | null
  min_merit_score: number | null
  income_ceiling: number | null
  per_capita_ceiling: number | null
}

/** One language's block of the gift's public apply-page copy. */
export interface AdminApplyCopyBlock {
  title: string
  intro: string
  criteria: string[]
}

/** The gift's copy as STORED — an absent locale means that box is blank, not English. */
export type AdminApplyCopy = Partial<Record<'en' | 'ms' | 'ta', AdminApplyCopyBlock>>

export interface AdminProgramme {
  id: number
  code: string
  name_en: string
  name_ms: string
  name_ta: string
  is_active: boolean
  /**
   * What the PUBLIC apply page says about this gift. `{}` = the platform's own wording.
   *
   * ⚠ THE STORED MAP, VERBATIM — the reader-facing endpoint folds ms/ta onto English, this does
   * NOT. The tab is an editor: a blank Malay box must render blank, or the first save would
   * silently promote the English text into a field nobody typed.
   */
  apply_copy: AdminApplyCopy
  /**
   * Race / ethnicity / religion words found in that copy. ADVISORY — nothing refuses on it.
   *
   * ⚠ `decisions.md` 2026-05-25 removed ethnicity from the public copy because MyNadi's s44(6)
   * tax status requires the programme not to discriminate by race. The owner ruled 2026-09-09
   * that this WARNS rather than refuses: an ethnicity-scoped gift is lawful in Malaysia and the
   * constraint is our funder's, not the platform's.
   */
  apply_copy_sensitive: string[]
  /**
   * Where the gift is in its life — the badge on its card.
   *
   * ⚠ SERVED, NOT DERIVED HERE (owner ruling, 2026-09-07). `draft` and `archived` are BOTH
   * `is_active: false`; the server splits them on whether anybody has ever applied. The card
   * carries an `applications` count, so this COULD be worked out in the browser — do not. It is
   * the same rule the Delete control reads, and two copies of it would put a **Draft** badge
   * beside a Delete button greyed because students applied.
   *
   * `is_active` stays beside it because it is what the PATCH writes; `lifecycle` is how it READS.
   */
  lifecycle: 'draft' | 'active' | 'archived'
  intake_years: number
  /**
   * Everyone who has ever applied to this gift.
   *
   * ⚠ COUNTED THROUGH THE ROUND AS WELL AS THE COLUMN, server-side. An application's gift is
   * denormalised and set once, so a round moved between gifts leaves its applications on the old
   * one — and this number has to agree with the Applications list, which narrows the same way.
   */
  applications: number
  /**
   * How many of them have EVER been awarded — not how many sit at `awarded` right now.
   *
   * ⚠ `awarded` is one stage in awarded → active → maintenance → closed, so a live status count
   * would FALL as students progress. Served from the never-cleared `awarded_at` stamp; do not
   * re-derive it here from anything on this card.
   */
  awarded: number
  /** The year currently taking applications, or null. A PROGRAMME is never open — a year is. */
  open_year: number | null
  /**
   * What is holding this gift, so the Delete control can be DISABLED with the reason showing —
   * or `null` when nothing is and it may be deleted.
   *
   * ⚠ SERVED, NEVER DERIVED HERE. This payload carries `intake_years` and `applications`; it has
   * never carried benefactors, money or payment runs. A button disabled on what the client happens
   * to know would go green for a gift held by a donation and refuse only after the phrase was
   * typed — rarer than the bug it replaces, and more surprising. The delete endpoint refuses from
   * the SAME function that fills this, so the two cannot disagree.
   *
   * ⚠ AN INTAKE YEAR IS NOT ON THIS LIST, and its absence is the owner's ruling (2026-09-07):
   * students hold a gift, a year on its own does not. An EMPTY year is deleted along with the
   * gift. Do not add `has_intake_years` back by reading `intake_years > 0` here — that is exactly
   * the derived-from-the-card mistake the paragraph above refuses.
   */
  delete_blocked_by:
    | 'has_applications' | 'has_benefactors'
    | 'has_money' | 'has_payment_runs' | null
  delete_blocked_count: number
  /**
   * The whole link a student follows to apply to THIS gift.
   *
   * ⚠ SERVED WHOLE — do not rebuild it from `code` and `window.location.origin`. The console and
   * the student site share an origin today, so that would be right, and would go silently wrong
   * the day a tenant is served from its own domain; the server already answers that per
   * organisation. It is also the one string somebody copies onto a poster.
   *
   * ⚠ PER GIFT, NOT PER YEAR (owner ruling). The code is the gift's permanent identifier, so a
   * printed link survives every intake; the server picks whichever round is open.
   */
  apply_url: string
}

export interface AdminIntakeYear {
  id: number
  code: string
  name: string
  year: number
  is_open: boolean
  is_active: boolean
  applications: number
  requirements: ProgrammeRequirements
  /** ⚠ THE ROUND'S STATED WINDOW, AND IT DESCRIBES — IT OPENS NOTHING (owner, 2026-09-06).
   *  `is_open` is the switch and stays the switch; these two say when the round is MEANT to run.
   *  `null` means no window was stated, which is a normal round — render a dash, not an error. */
  opens_on: string | null
  closes_on: string | null
  /**
   * Where the round is in its life — **SERVED, never derived here** (`views_admin.round_state`).
   *
   * ⚠⚠ FOUR STATES, THREE BEHAVIOURS, AND THE MIDDLE ONE IS LOAD-BEARING:
   *   · `open`     — anyone may start and submit.
   *   · `closed`   — no NEW applications; **anyone already started may still finish**. That grace
   *                  period is real: the 2026 round closed on 1 July and thirty students who were
   *                  part-way through submitted between then and the 7th.
   *   · `finished` — nobody may submit. **TERMINAL** — the server refuses to reopen one.
   *   · `draft`    — never opened, nobody has applied.
   */
  state: 'draft' | 'open' | 'closed' | 'finished'
  /** When the round was closed for good, and by whom. Null while it can still be reopened. */
  finished_at: string | null
  finished_by: string
  /** How many applicants have STARTED and not yet submitted — the people a finish would shut out.
   *  ⚠ Counted as `status='shortlisted'` server-side; `submitted_at` is `auto_now_add` and never
   *  null, so it cannot answer this. */
  unsubmitted: number
}

export async function getAdminProgrammes(options?: ApiOptions) {
  return adminFetch<{ programmes: AdminProgramme[] }>(
    '/api/v1/admin/scholarship/programmes/', options)
}

/** Create a gift. It arrives INACTIVE whatever is sent — switching it on is a separate press,
 *  because an active second programme changes live behaviour the moment it exists. */
export async function createAdminProgramme(
  body: { code: string; name_en: string; name_ms?: string; name_ta?: string },
  options?: ApiOptions,
) {
  return adminMutate<AdminProgramme>('/api/v1/admin/scholarship/programmes/', 'POST', body, options)
}

export async function updateAdminProgramme(
  id: number,
  // ⚠ `code` IS EDITABLE, AND THE SERVER KEEPS THE OLD ONE AS AN ALIAS. Sending it unchanged is a
  // no-op — no alias is written — so the rename dialog may post the box as it stands.
  // ⚠ `apply_copy` IS ALL-OR-NOTHING PER LANGUAGE and the SERVER validates it — length caps,
  // no markup, and "title + intro + a bullet, or none". Send `{}` to return the gift to the
  // platform's own wording.
  body: Partial<{
    code: string; name_en: string; name_ms: string; name_ta: string; is_active: boolean
    apply_copy: AdminApplyCopy
  }>,
  options?: ApiOptions,
) {
  return adminMutate<AdminProgramme>(
    `/api/v1/admin/scholarship/programmes/${id}/`, 'PATCH', body, options)
}

/**
 * Draft this gift's Malay or Tamil apply-page copy from ITS OWN saved English.
 *
 * ⚠⚠ IT RETURNS A DRAFT AND SAVES NOTHING. The caller puts the block in the boxes; the wording
 * only ever reaches a public page through `updateAdminProgramme`, pressed by a person. That is
 * the same division the document engines keep — the model proposes, a human decides.
 *
 * ⚠ IT READS THE SAVED ENGLISH, not what is on screen, so a caller must not offer it while the
 * English has unsaved edits: the draft would translate wording the reader can no longer see.
 * ⚠ It is BILLABLE — one model call per invocation.
 */
export async function draftApplyCopy(
  id: number, locale: 'ms' | 'ta', options?: ApiOptions,
): Promise<AdminApplyCopyBlock> {
  const r = await adminMutate<{ locale: string; block: AdminApplyCopyBlock }>(
    `/api/v1/admin/scholarship/programmes/${id}/apply-copy/draft/`, 'POST', { locale }, options)
  return r.block
}

/**
 * Delete a gift that never became anything.
 *
 * ⚠ `confirm` MUST BE THE GIFT'S OWN CODE, and the SERVER checks it — this is not a client
 * courtesy. A destructive verb any caller can fire with an empty body is one mis-wired button
 * away from deleting somebody's gift.
 *
 * ⚠ IT REFUSES WITH A NAMED REASON rather than a generic failure: `has_applications`,
 * `has_benefactors`, `has_money`, `has_payment_runs`. Those are the relations the model already
 * protects — a gift that has taken a student or a ringgit cannot be deleted.
 *
 * ⚠ THE GIFT'S EMPTY INTAKE YEARS ARE DELETED WITH IT (owner, 2026-09-07). A year is rules, not
 * students, so it never refuses; one that holds an application is caught as `has_applications`.
 */
export async function deleteAdminProgramme(
  id: number,
  confirm: string,
  options?: ApiOptions,
) {
  return adminMutate<void>(
    `/api/v1/admin/scholarship/programmes/${id}/`, 'DELETE', { confirm }, options)
}

export async function getAdminIntakeYears(programmeId: number, options?: ApiOptions) {
  return adminFetch<{
    programme: { id: number; code: string; name_en: string; is_active: boolean }
    years: AdminIntakeYear[]
  }>(`/api/v1/admin/scholarship/programmes/${programmeId}/years/`, options)
}

/** Create an intake year. It arrives CLOSED — opening is what lets real students in, and gets its
 *  own deliberate press. Requirement keys may be omitted; a `null` unticks that test. */
export async function createAdminIntakeYear(
  programmeId: number,
  body: { code: string; name: string; year: number; opens_on?: string | null; closes_on?: string | null }
    & Partial<ProgrammeRequirements>,
  options?: ApiOptions,
) {
  return adminMutate<AdminIntakeYear>(
    `/api/v1/admin/scholarship/programmes/${programmeId}/years/`, 'POST', body, options)
}

export async function updateAdminIntakeYear(
  id: number,
  body: Partial<{ name: string; is_open: boolean; opens_on: string | null; closes_on: string | null }>
    & Partial<ProgrammeRequirements>,
  options?: ApiOptions,
) {
  return adminMutate<AdminIntakeYear>(
    `/api/v1/admin/scholarship/intake-years/${id}/`, 'PATCH', body, options)
}

/**
 * Close a round FOR GOOD. **Terminal — nothing in the product undoes this.**
 *
 * ⚠ ITS OWN ENDPOINT, and `confirm` must be the round's own code. The server checks it; the typed
 * box is not the guard, it is the pause. The round must already be closed (`still_open` otherwise).
 */
export async function finishAdminIntakeYear(
  id: number, confirm: string, options?: ApiOptions,
) {
  return adminMutate<AdminIntakeYear>(
    `/api/v1/admin/scholarship/intake-years/${id}/finish/`, 'POST', { confirm }, options)
}

File: halatuju_api/apps/scholarship/models/programmes.py
"""
A programme, its code aliases, and an intake round (cohort).

Moved here VERBATIM from `apps/scholarship/models.py` at code health H15 (2026-09-20).
Moves only: no field, no `Meta`, no `db_table`, no `related_name` was touched, so the
database is untouched too — `makemigrations --check` reports no changes. See `__init__.py`.
"""
from django.db import models

class Programme(models.Model):
    """A gift programme — THE DURABLE LEVEL of the platform hierarchy (2026-07-26).

    Hierarchy: HalaTuju (platform) -> Organisation -> **Programme** -> Year (intake)
    -> the student's individual award.

    A Programme **IS** the gift ("the BrightPath Bursary", "the Sabah Bursary") — not a
    container for one. **One gift per programme** (owner ruling; see decisions.md
    "One gift per Programme", 2026-07-26). It is the level that NEVER LAPSES: students
    join annually into the same programme, and each annual intake is a
    ``ScholarshipCohort`` hanging beneath it.

    What lives where:
      * **Organisation** — branding, sender identity, staff, and THE SECURITY FENCE.
      * **Programme**    — the gift itself: its rules and (from a later sprint) its fund.
                           Rule DEFAULTS land here; today the tunables still live on the
                           cohort, which is why this model carries none yet.
      * **Year (cohort)** — the annual intake: open/closed, deadlines, per-intake overrides.

    This is NOT a second security boundary. The organisation fence
    (``_AdminBase._org_scoped`` / ``_org_allows``) is unchanged — programme is a
    narrowing INSIDE that wall, never a replacement for it.

    What the award is CALLED ("bursary" / "scholarship" / "assistance") is
    per-ORGANISATION wording resolved through ``branding.py`` — never a property of this
    model and never a behavioural switch. Every award is a GIFT, never a loan
    (platform invariant, decisions.md 2026-07-26).
    """
    organisation = models.ForeignKey(
        'courses.PartnerOrganisation', on_delete=models.PROTECT,
        related_name='programmes',
        help_text='The tenant organisation that runs this gift programme.',
    )
    code = models.CharField(
        max_length=50, unique=True,
        help_text="URL-safe slug, e.g. 'brightpath-flagship', 'brightpath-sabah'",
    )
    # Trilingual display name, mirroring the organisation's branding columns. A blank
    # ms/ta falls back to _en at render time (the branding.py fallback convention).
    name_en = models.CharField(max_length=200)
    name_ms = models.CharField(max_length=200, blank=True, default='')
    name_ta = models.CharField(max_length=200, blank=True, default='')
    is_active = models.BooleanField(default=True)

    # ── What the PUBLIC apply page says about this gift (2026-09-09) ─────────────────────────
    #
    # ⚠ BLANK MEANS THE PLATFORM DEFAULT, and blank is the correct state for BrightPath. The
    # stored map holds ONLY what an organisation wrote — never a copied default, which rots the
    # day the platform's own wording moves (the `OrganisationConfiguration` rule).
    #
    # ⚠ JSON RATHER THAN NINE COLUMNS, and the reason is the bullets: the criteria list is
    # variable-length (the owner's ruling — a gift may advertise three conditions or five), so a
    # column model needs a JSON column for it anyway; splitting title/intro out would buy six
    # more migrations' worth of drift and nothing else.
    #
    # ⚠ IT IS NOT DERIVED FROM THE ROUND'S THRESHOLDS AND MUST NEVER BE. The advertised bar is
    # deliberately stricter than `shortlisting.evaluate()` — see `apply_copy.py`.
    apply_copy = models.JSONField(
        default=dict, blank=True,
        help_text='Public apply-page copy per language: '
                  '{"en": {"title": …, "intro": …, "criteria": […]}, "ms": {…}, "ta": {…}}. '
                  'Blank means the platform default. Validated by apply_copy.normalise.',
    )
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'scholarship_programmes'
        ordering = ['organisation_id', 'code']

    def __str__(self):
        return f'{self.name_en} ({self.code})'


class ProgrammeCodeAlias(models.Model):
    """A code this gift USED to answer to — so a printed apply link never dies.

    ⚠⚠ THIS TABLE IS THE WHOLE REASON A GIFT CODE MAY BE RENAMED AT ALL.
    `Programme.code` is what `/scholarship/apply?p=<code>` carries, and `resolve_open_cohort`
    narrows on it. An unknown code narrows to NOTHING, which the apply page reads as **"no open
    round"** — so without an alias a rename would make every poster, WhatsApp forward and printed
    flyer already in circulation tell a student *"applications are closed"*, silently, with no
    error anywhere for us to see. The student simply goes away.

    ⚠ THE ALIAS SERVES THE STUDENT PATH ONLY (owner ruling, 2026-09-09). `resolve_open_cohort` is
    the one resolver that consults it. The admin console's own gift switcher
    (`_AdminBase._programme_by_code`) keeps resolving LIVE codes only: an admin's URL is never
    printed on a poster, and letting a stale code work there would hide a rename from the very
    people who performed it.

    ⚠ ONE WRITER, AND NAMING IT IS PART OF THE DESIGN. Rows are created by
    `AdminProgrammeDetailView.patch` at the moment a code changes, and by nothing else. A table
    populated by a backfill with no matching write path is a bug with a delay on it — migration
    `0123` left the 28/07 sponsor with no membership that way (lessons.md, 2026-07-29). There is
    no backfill here BECAUSE no code has ever been renamed; the first rename writes the first row.

    ⚠ UNIQUENESS SPANS BOTH TABLES, AND THE DATABASE CANNOT SAY SO. `code` is unique here, and
    `Programme.code` is unique there, but nothing stops an alias colliding with another gift's LIVE
    code — which would make one link resolve two ways. `code_is_free()` is the one check both the
    create and the rename paths call; it is application-level on purpose, since a cross-table
    constraint would need a trigger this project has no precedent for.
    """
    programme = models.ForeignKey(
        Programme, on_delete=models.CASCADE, related_name='code_aliases',
        help_text='The gift this retired code still points at.',
    )
    code = models.CharField(
        max_length=50, unique=True,
        help_text="A code this gift used to answer to, e.g. 'brightpath-flagship'.",
    )
    created_at = models.DateTimeField(auto_now_add=True)
    created_by = models.CharField(
        max_length=254, blank=True, default='',
        help_text='Email of the admin who renamed the code, for the audit trail.',
    )

    class Meta:
        db_table = 'scholarship_programme_code_aliases'
        ordering = ['programme_id', 'code']

    def __str__(self):
        return f'{self.code} -> {self.programme.code}'


def code_is_free(code, *, exclude_programme=None):
    """Is `code` available — as a live gift code AND as a retired one?

    ⚠ BOTH TABLES, ALWAYS. A code taken by another gift's ALIAS is not free: reusing it would make
    one printed link resolve to two different gifts depending on which query ran first. The two
    unique constraints each guard their own table and neither can see the other, so this function
    is the only place the real rule exists — call it from every path that accepts a code.

    `exclude_programme` lets a gift keep its own code (a no-op rename) and re-claim a code it had
    itself retired earlier, which is the natural "undo a rename" and must not be refused.
    """
    live = Programme.objects.filter(code=code)
    alias = ProgrammeCodeAlias.objects.filter(code=code)
    if exclude_programme is not None:
        live = live.exclude(pk=exclude_programme.pk)
        alias = alias.exclude(programme=exclude_programme)
    return not live.exists() and not alias.exists()


class ScholarshipCohort(models.Model):
    """
    A single application round — the YEAR (intake) level of the hierarchy: one annual
    round of students entering a ``Programme``. Holds the configurable shortlisting
    thresholds and funding parameters so they can be tuned without code changes (the
    shortlisting rules engine in Sprint 3 reads these).

    NOTE (2026-07-26): this model historically did TWO jobs — the programme (rules,
    funding envelope, eligibility) AND the intake year (``b40-2026``, ``year=2026``).
    ``Programme`` above now owns the durable half. Moving the tunables up to become
    programme-level DEFAULTS with per-intake overrides is deliberately a LATER sprint —
    those columns feed the verification engine, so that change is behaviour-sensitive
    and is kept out of the structural one.
    """
    # ── Tenant ownership (platform Sprint 1) ──────────────────────────────────
    # SOURCE OF TRUTH for "which organisation owns this programme". The
    # application-level denormalised copy (platform Sprint 2) must always equal
    # this — never store a second independently-mutable copy. Named
    # owning_organisation deliberately: `PartnerAdmin.org` / `referred_by_org`
    # mean the REFERRING org (attribution), never ownership/access control.
    # Nullable only for additive-migration safety; seeded to BrightPath (org #1)
    # by migration 0098 — NULL carries no meaning and nothing reads it yet.
    owning_organisation = models.ForeignKey(
        'courses.PartnerOrganisation', on_delete=models.PROTECT,
        null=True, blank=True, related_name='owned_cohorts',
        help_text='The tenant organisation that OWNS this programme (platform Sprint 1). '
                  'SOURCE OF TRUTH for tenancy. ScholarshipApplication.owning_organisation is '
                  'a denormalised copy of this, set in the application save(). A future '
                  '"move a cohort between organisations" flow MUST cascade the new value to '
                  'every one of that cohort.applications.owning_organisation (there is no DB '
                  'trigger — the drift guard test asserts they agree).',
    )
    # Platform programme layer (2026-07-26): the durable gift this intake belongs to.
    # ``organisation`` above stays the SOURCE OF TRUTH for tenancy/security; this is the
    # funding + rules level beneath it. Nullable for additive-migration safety and for
    # bare test fixtures; prod is backfilled. programme.organisation must agree with
    # owning_organisation — the drift guard test asserts it.
    programme = models.ForeignKey(
        'Programme', on_delete=models.PROTECT,
        null=True, blank=True, related_name='cohorts',
        help_text='The gift programme this annual intake belongs to (platform '
                  'programme layer). The programme never lapses; intakes cycle.',
    )

    code = models.CharField(
        max_length=50, unique=True,
        help_text="URL-safe slug, e.g. 'b40-2026'",
    )
    name = models.CharField(
        max_length=200,
        help_text="Display name, e.g. 'BrightPath Bursary Programme 2026'",
    )
    year = models.IntegerField()
    is_active = models.BooleanField(default=True)
    is_open = models.BooleanField(
        default=True, help_text="Currently accepting new applications",
    )

    # ── The round's stated window (owner, 2026-09-06) ────────────────────────────────────────
    #
    # ⚠⚠ THESE DATES DESCRIBE. THEY DO NOT OPEN OR CLOSE ANYTHING. `is_open` is the switch and
    # stays the switch; nothing reads these columns to decide whether a student may apply, and
    # nothing may be written that does. Owner ruling, 2026-09-06, taken against the alternative
    # of a scheduled job that opens on the date.
    #
    # Two reasons, and the second is the durable one:
    #   1. `is_open` already means "real students can walk in" (decisions.md, Sabah S2b). A date
    #      that ALSO opened would be a second switch that can disagree with the first, which is
    #      the exact shape `lessons.md` warns about — except here the switch already exists, so
    #      the date must not become a rival to it.
    #   2. A clock can fire BEFORE SETUP IS FINISHED. Create a gift, type a start date, get
    #      interrupted, and on that date the round opens with no rules set and no questions
    #      configured. A person pressing Open cannot do that by accident.
    #
    # ⚠ NULL IS A REAL ANSWER, AND NOTHING WAS BACKFILLED. A round with no stated window is a
    # normal round, not a broken one — every row that existed before this column (the live 2026
    # intake among them) has NULL, and inventing dates for a round that already ran would be
    # fiction on an audited row. Any reader must treat blank as "not stated", never as an error.
    #
    # If a timer is ever wanted, it is a job built ON TOP of these columns and argued on its own
    # merits then; it is not a reinterpretation of them.
    opens_on = models.DateField(
        null=True, blank=True,
        help_text='The date this round is STATED to open. Descriptive only — it opens nothing; '
                  '`is_open` is the switch. NULL means no window was stated.',
    )
    closes_on = models.DateField(
        null=True, blank=True,
        help_text='The date this round is STATED to close. Descriptive only — it closes nothing. '
                  'NULL means no window was stated.',
    )

    # ── Finishing a round for good (owner, 2026-09-08) ───────────────────────────────────────
    #
    # ⚠⚠ CLOSED AND FINISHED ARE DIFFERENT THINGS, AND THE DIFFERENCE IS A REAL BEHAVIOUR THAT
    # PRODUCTION ALREADY RELIED ON. `is_open=False` stops NEW applications and nothing else —
    # a student who had already started keeps their right to finish, because the intake gate
    # lives on the CREATE endpoint and a returning applicant never reaches it again. That is
    # stated at `views.ApplicationCreateView` and it is not an oversight.
    #
    # It is how the 2026 intake actually ran: the switch went off on 1 July, the landing page
    # greyed out, and THIRTY students who were already part-way through submitted between then
    # and 7 July. Nobody designed that grace period into the screen; it fell out of where the
    # gate sits, and for two months nothing said it existed. Reconstructing it needed a database
    # query, because there is no record of the close (this pair is the fix for that too).
    #
    # `finished_at` is the end of the grace period: the round is done, and a late submission is
    # refused. **It is TERMINAL** (owner: *"when an application is finished, can it be opened
    # again? I don't think it should be"*). Nothing in the product clears it — not the open
    # toggle, not the edit dialog. That is why finishing asks for the round's code to be TYPED,
    # the same shape as deleting a gift: an irreversible act gets a deliberate one.
    #
    # ⚠ DO NOT "SIMPLIFY" THIS INTO `is_open`. Three states are needed because the middle one is
    # load-bearing: open (anyone may start), closed (no new starts, those in flight may finish),
    # finished (nobody may submit). Collapsing closed into finished would have shut out those
    # thirty students on 1 July.
    finished_at = models.DateTimeField(
        null=True, blank=True,
        help_text='When this round was closed FOR GOOD. Terminal — nothing in the product '
                  'clears it. NULL means the round can still be reopened.',
    )
    finished_by = models.CharField(
        max_length=254, blank=True, default='',
        help_text="Email of the administrator who finished the round. Blank for a round "
                  "finished before this was recorded, which is never 'nobody'.",
    )

    # ── Shortlisting requirements (consumed by `shortlisting.evaluate`) ──────────────────────
    #
    # ⚠⚠ NULL MEANS THE TEST IS NOT APPLIED (Sabah S2a, owner 2026-09-02). Every one of these was
    # NOT NULL with a default, so every test always ran and an organisation had no way to say
    # "we do not use this one". BrightPath never asked for an STPM floor, and PNGK >= 2.90 applied
    # to all nine of its STPM applicants for a whole intake regardless.
    #
    # ⚠ THE VALUE **IS** THE SWITCH — there is deliberately no companion `use_x` boolean. Two
    # columns can disagree ("on, but blank"; "off, but 4") and then something has to decide which
    # wins, silently. One column cannot. The admin screen ticks a box by writing a value and
    # unticks it by clearing one.
    #
    # Defaults are kept so that existing rows and every test fixture behave exactly as before —
    # a NEW cohort still arrives with the flagship's floor and is edited from there.
    min_spm_a_count = models.IntegerField(
        null=True, blank=True, default=4,
        help_text="Minimum SPM grades at A- or better (A+/A/A- all count). NULL = not applied.",
    )
    min_spm_bplus_count = models.IntegerField(
        null=True, blank=True, default=5,
        help_text="Minimum SPM grades at B+ or better — the TOTAL strong count, not the extra "
                  "beyond the A's (4 A- plus 1 more B+ is stored as 5). NULL = not applied.",
    )
    min_stpm_pngk = models.FloatField(
        null=True, blank=True, default=2.9,
        help_text="Minimum STPM PNGK. NULL = not applied.",
    )
    min_merit_score = models.FloatField(
        null=True, blank=True, default=None,
        help_text="Minimum UPU merit point out of 100 (grades + co-curriculum). SPM applicants "
                  "only — an STPM applicant's comparable figure is their PNGK, which is "
                  "min_stpm_pngk. NULL = not applied, which is every cohort today.",
    )
    income_ceiling = models.IntegerField(
        null=True, blank=True,
        help_text="B40 monthly household GROSS income ceiling in RM (DOSM B40 line, RM5,860 in 2024). "
                  "PRIMARY income gate: a non-STR applicant at or below this passes regardless of household size. "
                  "NULL together with per_capita_ceiling = the financial test is not applied.",
    )
    per_capita_ceiling = models.IntegerField(
        null=True, blank=True, default=1584,
        help_text="Per-capita monthly income ceiling in RM (household_income / household_size). "
                  "SAFETY NET only — applies to non-STR applicants whose gross income is ABOVE income_ceiling, "
                  "rescuing large households. RM5,860 B40 ceiling / 3.7 avg household = RM1,584 (DOSM 2024). "
                  "NULL = this rescue is not offered.",
    )
    bucket_b_margin = models.IntegerField(
        default=1,
        help_text="DEPRECATED (pre-S8 marginal-miss logic); unused by the current engine",
    )

    # Funding + workflow parameters (consumed by later sprints)
    funding_envelope = models.IntegerField(
        null=True, blank=True, help_text="Per-student funding envelope in RM",
    )
    fail_email_delay_days = models.IntegerField(
        default=3,
        help_text="DEPRECATED (pre-S8); the scheduler now uses success/decline_delay_hours",
    )
    success_delay_hours = models.FloatField(
        default=48,
        help_text="Hours after submit before the shortlist (invitation) email + follow-up unlock (S8 delayed "
                  "reveal). Float so sub-hour delays are possible (e.g. 0.9167 = 55 minutes).",
    )
    decline_delay_hours = models.FloatField(
        default=48,
        help_text="Hours after submit before the warm decline email (S8 delayed reveal). Float (see above).",
    )
    # Check 2 STEP 2/3: days a student has to answer the AI clarify queries after submit
    # before the application proceeds to a reviewer regardless (the SLA clock, design §5).
    query_response_sla_days = models.PositiveSmallIntegerField(
        default=5,
        help_text="Check-2 query SLA: days after submit to answer clarify queries before "
                  "the application is ready for assignment regardless (proceed-as-is, flagged).",
    )

    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        db_table = 'scholarship_cohorts'
        ordering = ['-year', 'code']

    def __str__(self):
        return f'{self.name} ({self.code})'
